esp-hal/src/i2c.rs

```diff
@@ -240,11 +252,110 @@ pub struct I2c<'d, DM: Mode, T = AnyI2c> {
     timeout: Option<u32>,
 }
 
-impl<DM, T> I2c<'_, DM, T>
+impl<T> embedded_hal_02::blocking::i2c::Read for I2c<'_, Blocking, T>
+where
+    T: Instance,
+{
+    type Error = Error;
+
+    fn read(&mut self, address: u8, buffer: &mut [u8]) -> Result<(), Self::Error> {
+        self.read(address, buffer)
+    }
+}
+
+impl<T> embedded_hal_02::blocking::i2c::Write for I2c<'_, Blocking, T>
+where
+    T: Instance,
+{
+    type Error = Error;
+
+    fn write(&mut self, addr: u8, bytes: &[u8]) -> Result<(), Self::Error> {
+        self.write(addr, bytes)
+    }
+}
+
+impl<T> embedded_hal_02::blocking::i2c::WriteRead for I2c<'_, Blocking, T>
+where
+    T: Instance,
+{
+    type Error = Error;
+
+    fn write_read(
+        &mut self,
+        address: u8,
+        bytes: &[u8],
+        buffer: &mut [u8],
+    ) -> Result<(), Self::Error> {
+        self.write_read(address, bytes, buffer)
+    }
+}
+
+impl<T, DM: Mode> embedded_hal::i2c::ErrorType for I2c<'_, DM, T> {
+    type Error = Error;
+}
+
+impl<T, DM: Mode> embedded_hal::i2c::I2c for I2c<'_, DM, T>
+where
+    T: Instance,
+{
+    fn transaction(
+        &mut self,
+        address: u8,
+        operations: &mut [embedded_hal::i2c::Operation<'_>],
+    ) -> Result<(), Self::Error> {
+        self.transaction_impl(address, operations.iter_mut().map(Operation::from))
+    }
+}
+
+impl<'d, T, DM: Mode> I2c<'d, DM, T>
 where
     T: Instance,
-    DM: Mode,
 {
+    fn new_internal(
+        i2c: impl Peripheral<P = T> + 'd,
+        sda: impl Peripheral<P = impl PeripheralOutput> + 'd,
+        scl: impl Peripheral<P = impl PeripheralOutput> + 'd,
+        frequency: HertzU32,
+    ) -> Self {
+        crate::into_ref!(i2c);
+        crate::into_mapped_ref!(sda, scl);
+
+        let i2c = I2c {
+            i2c,
+            phantom: PhantomData,
+            frequency,
+            timeout: None,
+        };
+
+        PeripheralClockControl::reset(i2c.i2c.peripheral());
+        PeripheralClockControl::enable(i2c.i2c.peripheral());
+
+        let i2c = i2c.with_sda(sda).with_scl(scl);
+
+        i2c.info().setup(frequency, None);
```

Comment:
I need to check if we can flip the order here. If so, we can a) reuse internal_recover, and b) remove the pin params from the constructor.